util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/AttributesUtil.java

```diff
@@ -0,0 +1,229 @@
+package com.google.cloud.healthcare.imaging.dicomadapter;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.dcm4che3.data.Attributes;
+import org.dcm4che3.data.Sequence;
+import org.dcm4che3.data.Tag;
+import org.dcm4che3.data.VR;
+import org.dcm4che3.net.Status;
+import org.dcm4che3.net.service.DicomServiceException;
+import org.dcm4che3.util.TagUtils;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class AttributesUtil {
+  private static final String PN_ALPHABETIC = "Alphabetic";
+  private static final String PN_IDEOGRAPHIC = "Ideographic";
+  private static final String PN_PHONETIC = "Phonetic";
+  private static final String PN_DELIMITER = "=";
+
+  public static String getTagValue(JSONObject json, String tag) throws JSONException {
+    JSONObject jsonTag = json.getJSONObject(tag);
+    JSONArray valueArray = jsonTag.getJSONArray("Value");
+    if (valueArray.length() != 1) {
+      throw new JSONException(
+          "Expected single value for tag: "
+              + tag
+              + " in JSON:\n"
+              + valueArray.toString());
+    }
+    return valueArray.getString(0);
+  }
+
+  public static String getTagValueOrNull(JSONObject json, String tag) {
+    if (!json.has(tag)) {
+      return null;
+    }
+    JSONObject jsonTag = json.getJSONObject(tag);
+    if (!jsonTag.has("Value")) {
+      return null;
+    }
+    JSONArray valueArray = jsonTag.getJSONArray("Value");
+    if (valueArray.length() != 1) {
+      return null;
+    }
+    return valueArray.getString(0);
+  }
+
+  /**
+   * Returns array of QIDO-RS paths (more than 1 if multiple ModalitiesInStudy are present)
+   * @param attrs dcm4che Attributes to convert
+   * @param includeFields additonal includeFields for QIDO-RS
+   */
+  public static String[] attributesToQidoPathArray(Attributes attrs, String... includeFields)
+      throws IOException {
+    if (!attrs.containsValue(Tag.ModalitiesInStudy)
+        || attrs.getStrings(Tag.ModalitiesInStudy).length <= 1) {
+      return new String[]{attributesToQidoPath(attrs, includeFields)};
+    }
+
+    List<String> result = new ArrayList<>();
+    for (String modality : attrs.getStrings(Tag.ModalitiesInStudy)) {
+      Attributes attrsCopy = new Attributes(attrs);
+      attrsCopy.setString(Tag.ModalitiesInStudy, VR.CS, modality);
+      result.add(attributesToQidoPath(attrsCopy, includeFields));
+    }
+    return result.toArray(new String[]{});
+  }
+
+  /**
+   * Returns corresponding QIDO-RS path
+   * @param attrs dcm4che Attributes to convert
+   * @param includeFields additonal includeFields for QIDO-RS
+   */
+  public static String attributesToQidoPath(Attributes attrs, String... includeFields)
```

Comment:
Iterating code doesn't try to use comma notation, it sends tag1=[val1,val2,val3] as tag1=val1&tag1=val2&tag1=val3, which is a syntax by api (and server should support UID list matching according to standard, but currently doesn't).

Only UID lists are part of standard, so I could throw error when parameter VR is not UID list, but didn't see much meaning to doing so (since it's easy to handle server-side by ignoring parameter mentions past 1st or throwing error like it does now). In theory api could later choose to support lists for some other parameters too, even if it's not in standard.

Multi-value ModalitiesInStudy was used in clients commonly enough  (despite not being part of standard) to warrant special case handling.

Reply:
Ok will throw exception. But I think it's overly restrictive: even if api doesn't currently support this, it returns sane error which leads to same end result (Dimse Processing Failure).